driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionKmsTlsTest.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import com.mongodb.ClientEncryptionSettings;
+import com.mongodb.MongoClientException;
+import com.mongodb.client.model.vault.DataKeyOptions;
+import com.mongodb.client.vault.ClientEncryption;
+import org.bson.BsonDocument;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateExpiredException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.mongodb.client.Fixture.getMongoClientSettings;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.jupiter.api.Assumptions.assumeTrue;
+
+public abstract class AbstractClientSideEncryptionKmsTlsTest {
+
+    enum TlsErrorType {
+        EXPIRED() {
+            @Override
+            Class<?> getExpectedExceptionClass() {
+                return CertificateExpiredException.class;
+            }
+
+            @Override
+            String getExpectedExcepionMessageSubstring() {
+                return "NotAfter";
+            }
+        },
+        INVALID_HOSTNAME() {
+            @Override
+            Class<?> getExpectedExceptionClass() {
+                return CertificateException.class;
+            }
+
+            @Override
+            String getExpectedExcepionMessageSubstring() {
+                return "No subject alternative names present";
+            }
+        };
+        abstract Class<?> getExpectedExceptionClass();
+
+        abstract String getExpectedExcepionMessageSubstring();
+    }
```

Comment:
And then, we can go further, remove methods `containsCauseOfClass`, `getCauseOfClass` and change `TlsErrorType` such that  its values are objects (as in OOP) rather than structures:
```java
enum TlsError {
    EXPIRED(CertificateExpiredException.class, "NotAfter"),
    INVALID_HOSTNAME(CertificateException.class, "No subject alternative names present");

    private final Class<? extends Exception> expectedExceptionClass;
    private final String expectedExceptionMessageSubstring;

    TlsErrorType(final Class<? extends Exception> expectedExceptionClass, String expectedExceptionMessageSubstring) {
        this.expectedExceptionClass = expectedExceptionClass;
        this.expectedExceptionMessageSubstring = expectedExceptionMessageSubstring;
    }

    void assertMatches(final MongoClientException e) {
        Throwable found = e;
        while (found != null && !found.getClass().equals(expectedExceptionClass)) {
            found = cause.getCause();
        }
        assertNotNull(found);
        assertTrue(found.getMessage() != null && found.getMessage().contains(expectedExceptionMessageSubstring));
    }
}
```

With the above change, the assertions on this test become
```java
EXPECTED_KMS_TLS_ERROR.assertMatches(e)
```

Reply:
I like the first suggestion, but the second not so much.  I like to keep assertions in the test method itself and not delegated to another method unless there's a good reason.  Removing a couple of simple methods is not a good enough reason for me.

Reply:
Since the methods `containsCauseOfClass`, `getCauseOfClass` are staying, may be make them static and remove code duplication? Just as an example (another example was `assertMatches`), `containsCauseOfClass` can be implemented via `getCauseOfClass`, provided that `getCauseOfClass` returns nullable values instead of throwing `IllegalStateException`.

Reply:
I split the difference: I moved all the logic into the enum, but left the actual assertions in the test method.

Reply:
Done, I hope.